File: kotlin/exercises.kt

fun main(vararg args: String) {
	println(helloWorld())
	println(alternate("Hlo ol!", "el,Wrd"))
	println(String(swapEncrypt("Hello World".toCharArray(), 11)))
	println(String(swapEncrypt("eHll ooWlrd".toCharArray(), 11)))
	println(isPalindrome("Hello, World!"))
	println(isPalindrome("A Santa dog lived as a devil God at NASA."))
	for (i in 0..10)
		println(fibonacci(i))
}

fun helloWorld() = "Hello, World!"

fun alternate(first: String, second: String) : String {
	var result = ""
	val I = Math.min(first.length, second.length)
	for (i in 0 .. I - 1)
		result += first.substring(i, i + 1) + second.substring(i, i + 1)
	result += first.substring(I) + second.substring(I)
	return result
}

fun swapEncrypt(toSwap: CharArray, length: Int) : CharArray {
	for (i in 0 .. length - 2 step 2) {
		val tmp = toSwap[i]
		toSwap[i] = toSwap[i + 1]
		toSwap[i + 1] = tmp 
	}
	return toSwap
}

fun isPalindrome(palindrome: String) : Boolean {
	val clean = java.util.regex.Pattern.compile("[^A-Za-z]").matcher(palindrome).replaceAll("").toLowerCase()
	var i = 0; var j = clean.length - 1
	while (i < j) {
		if (clean[i] != clean[j]) return false
		i++; j--
	}
	return true
}

fun fibonacci(num: Int) : Int {
	if (num < 1) return 0
	if (num == 1) return 1
	return fibonacci(num - 2) + fibonacci(num - 1)
}

File: python/exercises.py
import re

def helloWorld(): return "Hello, World!"

def alternate(first, second):
	result = ''
	I = min(len(first), len(second))
	for i in range(I):
		result += first[i] + second[i]
	result += first[I:] + second[I:]
	return result

def swapEncrypt(toSwap, length):
	for i in range(0, length - 1, 2):
		tmp = toSwap[i]
		toSwap[i] = toSwap[i + 1]
		toSwap[i + 1] = tmp
	return toSwap

def isPalindrome(palindrome):
	clean = re.sub('[^A-Za-z]', '', palindrome).lower()
	i = 0; j = len(clean) - 1
	while i < j:
		if (clean[i] != clean[j]): return False
		i += 1; j -= 1
	return True

def fibonacci(num):
	if (num < 1): return 0
	if (num == 1): return 1
	return fibonacci(num - 2) + fibonacci(num - 1)

print(helloWorld())
print(alternate('Hlo ol!', 'el,Wrd'))
print(''.join(swapEncrypt(list('Hello World'), 11)))
print(''.join(swapEncrypt(list('eHll ooWlrd'), 11)))
print(isPalindrome('Hello, World!'))
print(isPalindrome('A Santa dog lived as a devil God at NASA.'))
for i in range(10):
	print(fibonacci(i))
